### PR TITLE
Optimize timestamp parsing

### PR DESCRIPTION
**Pull request by vojtechtoman**
# Which issue does this PR close?

It is related to the optimization work done in the scope of https://github.com/apache/arrow-datafusion/issues/9090.

# Rationale for this change

While investigating https://github.com/apache/arrow-datafusion/issues/9090 and profiling the bottlenecks in the `to_timestamp` implementation, it turned out that there was potential for improvements in the underlying `string_to_datetime` function.

# What changes are included in this PR?

* `TimestampParser: eliminate unnecessary calls to `wrapping_sub`
* `string_to_datetime`: introduce a constant instead of the expensive call `NaiveTime::from_hms(0, 0, 0)`
* timezone `parse_fixed_offset`: reorganize the code to save some unnecessary computation

The `parse_timestamp` benchmarks show an improvement in the range of 5-10% (on my machine) in some of the cases.

# Are there any user-facing changes?

No

### CODE REVIEW

File: arrow-cast/src/parse.rs [hunk 1]
```diff
@@ -24,9 +24,18 @@ use chrono::prelude::*;
...
+    digits[..N]
+        .iter()
+        .fold(0_u32, |acc, v| acc * 10 + *v as u32)
+        * 10_u32.pow((9 - N) as _)
+}
+
 /// Parse nanoseconds from the first `N` values in digits, subtracting the offset `O`
 #[inline]
-fn parse_nanos<const N: usize, const O: u8>(digits: &[u8]) -> u32 {
+fn parse_nanos_subtracting<const N: usize, const O: u8>(digits: &[u8]) -> u32 {
```
Comment (tustvold) [inline arrow-cast/src/parse.rs:38]: Have you benchmarked to confirm that this separation makes a difference to performance, I would be extremely surprised if these don't compile to the same thing

File: arrow-cast/src/parse.rs [hunk 2]
```diff
@@ -171,25 +181,30 @@ impl TimestampParser {
 ///
 /// [IANA timezones]: https://www.iana.org/time-zones
 pub fn string_to_datetime<T: TimeZone>(timezone: &T, s: &str) -> Result<DateTime<T>, ArrowError> {
+    #[allow(deprecated)]
```
Comment (tustvold) [inline arrow-cast/src/parse.rs:184]: I'm pretty lukewarm on this as I understand chrono intends to remove these methods in future. Is there a NaiveTime::MIN or something we could use?

File: arrow-cast/src/parse.rs [hunk 3]
```diff
@@ -171,25 +181,30 @@ impl TimestampParser {
...
+    const ZERO_TIME: NaiveTime = NaiveTime::from_hms(0, 0, 0);
+
     let err =
         |ctx: &str| ArrowError::ParseError(format!("Error parsing timestamp from '{s}': {ctx}"));
 
     let bytes = s.as_bytes();
-    if bytes.len() < 10 {
+    let len = bytes.len();
```
Comment (tustvold) [inline arrow-cast/src/parse.rs:191]: Again I am very surprised if this makes any performance difference, compilers are smart enough to figure this out

File: arrow-array/src/timezone.rs [hunk 4]
```diff
@@ -25,27 +25,43 @@ pub use private::{Tz, TzOffset};
 fn parse_fixed_offset(tz: &str) -> Option<FixedOffset> {
```
Comment (tustvold) [inline arrow-array/src/timezone.rs:25]: My benchmarks are not showing this code as being made any faster (or slower)

**Review by tustvold — COMMENTED** [review]
Thank you for this, always interested in ways to make the code faster.

That being said running the benchmarks on a cloud server with `RUSTFLAGS="-C target-cpu=native"` I get the following

```
2020-09-08              time:   [31.217 ns 31.223 ns 31.230 ns]
                        change: [-0.2267% +0.1475% +0.4724%] (p = 0.49 > 0.05)
                        No change in performance detected.
Found 11 outliers among 100 measurements (11.00%)
  6 (6.00%) high mild
  5 (5.00%) high severe

2020-09-08T13:42:29     time:   [43.968 ns 43.981 ns 43.994 ns]
                        change: [-2.4837% -2.1841% -1.9688%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

2020-09-08T13:42:29.190 time:   [47.942 ns 47.961 ns 47.988 ns]
                        change: [-1.7998% -1.4780% -1.1860%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild
  3 (3.00%) high severe

2020-09-08T13:42:29.190855
                        time:   [45.677 ns 45.700 ns 45.732 ns]
                        change: [-2.6798% -2.3640% -2.0804%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  2 (2.00%) low mild
  4 (4.00%) high mild
  4 (4.00%) high severe

2020-09-08T13:42:29.190855999
                        time:   [47.943 ns 47.982 ns 48.054 ns]
                        change: [-2.4814% -2.1780% -1.8673%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) high mild
  3 (3.00%) high severe

2020-09-08T13:42:29+00:00
                        time:   [52.647 ns 52.659 ns 52.674 ns]
                        change: [-0.7667% -0.6271% -0.4011%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 14 outliers among 100 measurements (14.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  10 (10.00%) high severe

2020-09-08T13:42:29.190+00:00
                        time:   [56.727 ns 56.745 ns 56.766 ns]
                        change: [-0.3171% -0.0705% +0.0792%] (p = 0.71 > 0.05)
                        No change in performance detected.
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) high mild
  3 (3.00%) high severe

2020-09-08T13:42:29.190855+00:00
                        time:   [54.198 ns 54.220 ns 54.260 ns]
                        change: [-0.7361% -0.4207% -0.1210%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 8 outliers among 100 measurements (8.00%)
  4 (4.00%) high mild
  4 (4.00%) high severe

2020-09-08T13:42:29.190855999-05:00
                        time:   [54.396 ns 54.404 ns 54.414 ns]
                        change: [-0.5175% -0.2003% +0.1082%] (p = 0.24 > 0.05)
                        No change in performance detected.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  3 (3.00%) high severe

2020-09-08T13:42:29.190855Z
                        time:   [44.025 ns 44.036 ns 44.055 ns]
                        change: [-3.1903% -2.9539% -2.8174%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  3 (3.00%) high mild
  6 (6.00%) high severe
```

So I can reproduce a very slight performance uplift, but I suspect there are a fair few changes mixed in here that are negatively impacting legibility for no performance uplift. Perhaps we could reduce this down to the changes that yield a benefit?

**Comment by vojtechtoman** [comment]
@tustvold Many thanks for the review. Let's scrap this PR. I did benchmark my changes locally and even though the results were rather volatile, I saw a small improvement. But I tend to agree with you that the gains are too unsubstantial to justify the changes.

**Comment by tustvold** [comment]
Ok, I may see if I can extract some of the changes over the next few days. A small performance improvement is still an improvement 😅

**Pull request closed without merging**